### PR TITLE
Reorganize wallets packages

### PR DESCRIPTION
**Pull request by jordansexton**
This is intended to fix https://github.com/solana-labs/wallet-adapter/issues/183. It may also help with https://github.com/solana-labs/wallet-adapter/issues/153.

The problem is that the `wallets` package contains an enum of all the wallet names that is required by the React packages and most other packages. This makes it impossible to add support for new wallets without modifying this package, and it also makes it impossible to use the React package without installing all the wallets, even ones that aren't used.

The core changes are 

- Move the `Wallet` type definition into the `base` package: https://github.com/solana-labs/wallet-adapter/blob/wallets-reorg/packages/core/base/src/wallet.ts, and remove the `WalletName` enum, and replace it with a branded nominal type. These changes allow any wallet package to provide a compatible `Wallet` to use.
- Change the type of the `adapter` field of the `Wallet` definition to `Adapter` rather than `() => Adapter`. This allows a single adapter instance to be reused instead of being instantiated every time the selected wallet changes.
- Move the corresponding `Wallet` getter to every wallet package, e.g. https://github.com/solana-labs/wallet-adapter/blob/wallets-reorg/packages/wallets/phantom/src/wallet.ts.
- Change the `ready` interface of the `WalletAdapter` and implement it in each wallet package, e.g. https://github.com/solana-labs/wallet-adapter/blob/wallets-reorg/packages/wallets/phantom/src/adapter.ts#L66-L79. This fixes the dependence on polling, and combined with the reused adapter instance, will enable detection of multiple ready wallets to make use of what the user has installed (use of this feature isn't covered in this PR).
- Moves the `wallets` package and refactors it to simply export the other wallet packages: https://github.com/solana-labs/wallet-adapter/blob/wallets-reorg/packages/wallets/wallets/src/index.ts
- Refactors the `react` package to use these new APIs: https://github.com/solana-labs/wallet-adapter/blob/wallets-reorg/packages/core/react/src/WalletProvider.tsx#L63-L83
- Partially refactors the [Angular](
https://github.com/solana-labs/wallet-adapter/blob/wallets-reorg/packages/core/angular/src/wallet/wallet.store.ts#L132), [Svelte](https://github.com/solana-labs/wallet-adapter/blob/wallets-reorg/packages/core/svelte/src/walletStore.ts#L282), and [Vue](https://github.com/solana-labs/wallet-adapter/blob/wallets-reorg/packages/core/vue/src/useWallet.ts#L128) packages to do the same, but this work is unfinished (see the links to each package).

@danmt @lorisleiva @silvestrevivo @cogoo I could really use your advice/help with this last item!

### CODE REVIEW

**Comment by zicklag** [comment]
I haven't tested or checked out the code, but the list of changes sounds like it will really help me by splitting things out. Thanks!

**Comment by lorisleiva** [comment]
Hi 👋 

I'm going to test this branch with my dummy Vue app to see if it breaks things.

Regarding "FIXME: Asynchronously update the ready state" in the Vue core package:

I would probably add the asynchronous update inside the `setStateFromAdapter` method since this is used twice within the `initWallet` method. The following should work.

```ts
const setStateFromAdapter = (wallet: Wallet, adapter: Adapter) => {
    setState({
        wallet,
        adapter,
        ready: false,
        publicKey: adapter.publicKey,
        connected: adapter.connected,
    });

    // Asynchronously update the ready state
    const waiting = name;
    (async function () {
        const newReady = await adapter.ready();
        // If the selected wallet hasn't changed while waiting, update the ready state
        if (name === waiting) {
            ready.value = newReady;
        }
    })();
};
```

File: packages/wallets/blocto/src/adapter.ts
Comment (jordansexton) [inline packages/wallets/blocto/src/adapter.ts:40]: Should be `!== 'undefined'`

**Review by vpontis — APPROVED** [review]
This is great thanks @jordansexton! I didn't follow all of the refactoring but I checked out the branch and tested the Next.js repo which seems to work great 👍. 

Also I wonder if it makes sense for you to actively maintain all of these different wallet connectors in this repo. It sure seems like a lot of work! 

Maybe when this gets more stable, it'll make sense to split off individual wallet connectors into their own services. Or maybe there is a way to have one package connect to many wallets — like a better version of WalletConnect...

**Comment by cogoo** [comment]
Looks good to me from the Angular / Svelte point of view. I'll have to update #173 once this is merged

**Comment by silvestrevivo** [comment]
@jordansexton same than @cogoo The refactor on #173 makes the store so different that we can wait till this is merged to make changes.

File: packages/wallets/blocto/src/adapter.ts
Comment (jordansexton) [reply packages/wallets/blocto/src/adapter.ts]: Fixed.

File: packages/core/base/src/wallet.ts
Comment (jordansexton) [reply packages/core/base/src/wallet.ts:4]: This is a bit of a trick to get nominal types: https://medium.com/@KevinBGreene/surviving-the-typescript-ecosystem-branding-and-type-tagging-6cf6e516523d

The idea here is that any wallet can declare that its name is a `WalletName` by casting it. Downstream of this, the React/Vue/whatever components still have strong(ish) typing so that any string isn't accepted to select a wallet. Userland code never needs to cast wallet names, and they still get (some) type safety.

**Comment by jordansexton** [comment]
> Looks good to me from the Svelte point of view. I'll have to update #173 once this is merged. The API for the Svelte store has changed significantly, so will prefer to make the relevant changes in that PR. Thoughts?

@cogoo sounds great. Since the Angular and Svelte packages aren't published anyway, I'll probably merge this today after some wallet testing.

File: packages/core/base/src/wallet.ts
Comment (vpontis) [reply packages/core/base/src/wallet.ts:4]: Interesting, makes sense! I might suggest using the `Brand` generic that the article explains and adding a link to the article in a comment. That might help other confused people in the future :)

**Comment by jordansexton** [comment]
> I would probably add the asynchronous update inside the `setStateFromAdapter` method since this is used twice within the `initWallet` method.

@lorisleiva I implemented this in 443d2a44b3d7c429f5e640591fcf3bf3c70a3679, please let what you think. It shouldn't be called on `connect` -- the state must be ready to connect anyuway -- so it only needs to be called when the adapter changes.

File: packages/core/base/src/wallet.ts
Comment (jordansexton) [reply packages/core/base/src/wallet.ts]: Done.

**Comment by jordansexton** [comment]
> EDIT: So I tried for about an hour to use this PR in a dummy Vue3 app but couldn't do it successfully. Normally I just need to symlink `vue` and `vue-ui` but this time everything had to be symlinked together and I got lots of weird errors along the way. Is there a way I can use the local packages in an app without using symlinks?

@lorisleiva Hmm, I don't know. I did fix the build just today though. If you run `yarn build` from the base directory, all the packages build now. But since the changes aren't published, I'm not sure how you'd reference it without symlinks. You could create your Vue3 app as a package inside the source maybe?

**Comment by jordansexton** [comment]
Merging this so that other PRs can catch up with the changes, but I won't publish until Vue is known to work. I need to test with some of the lesser used wallets too.

**Comment by lorisleiva** [comment]
Thanks @jordansexton, these changes look good!

I ended up having to compile and copy all the required `lib` folders inside the [test app](https://github.com/lorisleiva/solana-vue3-counter/tree/test/wallets-reorg) because otherwise `vue` kept complaining that two versions of the same component existed and failed to render anything.

But yeah **everything works fine for VueJS applications** on the latest commit of `master` so I think we're good to publish these packages. 🚀 

I'll make sure to update #192 accordingly as well.

**Pull request merged**
